pkg/syncer/cnsoperator/controller/virtualmachinesnapshot/virtualmachinesnapshot_controller.go

```diff
@@ -0,0 +1,403 @@
+// will remove the work from the queue.
+func (r *ReconcileVirtualMachineSnapshot) Reconcile(ctx context.Context,
+	request reconcile.Request) (reconcile.Result, error) {
+	log := logger.GetLogger(ctx)
+	log.Info("Reconciling CR VirtualMachineSnapshot")
+	// Fetch the VirtualMachineSnapshot instance.
+	vms := &vmoperatorv1alpha4.VirtualMachineSnapshot{}
+	err := r.client.Get(ctx, request.NamespacedName, vms)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Infof("VirtualMachineSnapshot resource not found. Ignoring since object must be deleted.")
+			return reconcile.Result{}, nil
+		}
+		log.Errorf("Error reading the VirtualMachineSnapshot with name: %q. Err: %+v",
+			request.Name, err)
+		// Error reading the object - return with err.
+		return reconcile.Result{}, err
+	}
+
+	// Initialize backOffDuration for the instance, if required.
+	backOffDurationMapMutex.Lock()
+	var timeout time.Duration
+	if _, exists := backOffDuration[request.NamespacedName]; !exists {
+		backOffDuration[request.NamespacedName] = time.Second
+	}
+	timeout = backOffDuration[request.NamespacedName]
+	backOffDurationMapMutex.Unlock()
+
+	log.Info("Reconciling virtualmachinesnapshot")
+	err = r.reconcileNormal(ctx, vms)
+	if err != nil {
+		return reconcile.Result{RequeueAfter: timeout}, err
+	}
+	backOffDurationMapMutex.Lock()
+	delete(backOffDuration, request.NamespacedName)
+	backOffDurationMapMutex.Unlock()
+	return reconcile.Result{}, nil
+}
+func (r *ReconcileVirtualMachineSnapshot) reconcileNormal(ctx context.Context,
+	vmsnapshot *vmoperatorv1alpha4.VirtualMachineSnapshot) error {
+	log := logger.GetLogger(ctx)
+	// Check for the annotation "csi.vsphere.volume.sync: Requested"
+	annotationValue := vmsnapshot.Annotations["csi.vsphere.volume.sync"]
+	if annotationValue == "Requested" {
+		// if found fetch vmsnapshot and pvcs and pvs
+		vmKey := types.NamespacedName{
+			Namespace: vmsnapshot.Namespace,
+			Name:      vmsnapshot.Spec.VMRef.Name,
+		}
+		virtualMachine, _, err := utils.GetVirtualMachineAllApiVersions(ctx, vmKey,
+			r.vmOperatorClient)
+		if err != nil {
+			log.Errorf("ReconcileVirtualMachineSnapshot: "+
+				" Could not get VirtualMachine with name: %s namespace: %s err: %v",
+				virtualMachine.Namespace, virtualMachine.Name, err)
+			return err
+		}
+		err = r.validateStorageQuotaforVMSnapshot(ctx, virtualMachine)
+		if err != nil {
+			log.Errorf("ReconcileVirtualMachineSnapshot: Failed to validate VirtualMachineSnapshot for %s/%s err: %v",
+				vmsnapshot.Namespace, vmsnapshot.Name, err)
+			return err
+		}
+		// Update VMSnapshot CR annotation to "csi.vsphere.volume.sync: Completed"
+		vmsnapshot.Annotations["csi.vsphere.volume.sync"] = "Completed"
+		err = utils.UpdateVirtualMachineSnapshot(ctx, r.vmOperatorClient, vmsnapshot)
+		if err != nil {
+			log.Errorf("ReconcileVirtualMachineSnapshot: Could not update VirtualMachineSnapshot CR for %s/%s err: %v",
+				vmsnapshot.Namespace, vmsnapshot.Name, err)
+			return err
+		}
```

Comment:
Same question, what if this fails, but, the cnsvolumeinfo patch was a success?

Reply:
it is idempotent, case for multiple updates with same value is already handled
ref: https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/syncer/metadatasyncer.go#L4280